File: apps/mobile/lib/data/ws/jws.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as crypto;
import 'package:pointycastle/export.dart';

/// ES256 (RFC 7518 §3.4) JWS Compact Serialization (RFC 7515 §3.1) — the
/// Dart mirror of internal/receipt/jws.go, sharing one correctness bar (RFC
/// 7515 Appendix A.3) with the Go side (MADR 0077 D7). No JWS library is
/// used on either side: `ECDSASigner`/`ECSignature` expose the raw `r`/`s`
/// `BigInt` values directly, the same raw-value API `client_identity.dart`
/// already relies on for this device's own key.

/// The fixed JWS Protected Header this module produces and accepts. No
/// "kid": device/daemon identity is established out-of-band via the
/// enrolled-cert lookup (MADR 0077 D9), not a JWS header field, which would
/// be one more place to spoof.
const _jwsHeader = '{"alg":"ES256"}';
final String _jwsHeaderB64 = _encodeB64Url(utf8.encode(_jwsHeader));

String _encodeB64Url(List<int> bytes) =>
    base64Url.encode(bytes).replaceAll('=', '');

/// base64Url.decode requires canonical padding; base64 also rejects the
/// standard alphabet's +/ here, which is what we want for a URL-safe field
/// (same idiom as pair_uri.dart's `_tryDecodeB64Url`).
Uint8List? _decodeB64Url(String s) {
  final padded = s.padRight((s.length + 3) & ~3, '=');
  try {
    return base64Url.decode(padded);
  } on FormatException {
    return null;
  }
}

/// Big-endian, left-zero-padded to exactly [length] bytes — the ES256
/// signature encoding RFC 7518 §3.4 requires (raw R||S, not ASN.1 DER).
/// Mirrors Go's `big.Int.FillBytes`.
void _writeFixedBigEndian(Uint8List out, int offset, int length, BigInt value) {
  var v = value;
  for (var i = length - 1; i >= 0; i--) {
    out[offset + i] = (v & BigInt.from(0xff)).toInt();
    v = v >> 8;
  }
}

BigInt _readBigEndian(Uint8List bytes) {
  var result = BigInt.zero;
  for (final b in bytes) {
    result = (result << 8) | BigInt.from(b);
  }
  return result;
}

/// A P-256 `SecureRandom` seeded from `dart:math`'s cryptographically secure
/// generator — pointycastle does not seed one for you.
SecureRandom _seededSecureRandom() {
  final rng = FortunaRandom();
  final seedSource = Random.secure();
  final seed = Uint8List.fromList(
    List<int>.generate(32, (_) => seedSource.nextInt(256)),
  );
  rng.seed(KeyParameter(seed));
  return rng;
}

/// Signs [payload] as a JWS in Compact Serialization using ES256, matching
/// `internal/receipt/jws.go`'s `SignES256Compact` byte for byte: the
/// signature is the raw 64-byte R||S concatenation, big-endian, 32 bytes
/// each — never ASN.1 DER.
String signEs256Compact(ECPrivateKey priv, Uint8List payload) {
  final payloadB64 = _encodeB64Url(payload);
  final signingInput = '$_jwsHeaderB64.$payloadB64';
  final digest = Uint8List.fromList(
    crypto.sha256.convert(utf8.encode(signingInput)).bytes,
  );

  final signer = ECDSASigner(null, null)
    ..init(
      true,
      ParametersWithRandom(PrivateKeyParameter(priv), _seededSecureRandom()),
    );
  final sig = signer.generateSignature(digest) as ECSignature;

  final sigBytes = Uint8List(64);
  _writeFixedBigEndian(sigBytes, 0, 32, sig.r);
  _writeFixedBigEndian(sigBytes, 32, 32, sig.s);

  return '$signingInput.${_encodeB64Url(sigBytes)}';
}

/// Verifies [compact] against [pub] and, on success, returns the decoded
/// payload bytes; `null` on any failure — malformed input, wrong part count,
/// an `alg` other than exactly `ES256` (no algorithm negotiation), a
/// wrong-length signature, or a signature that does not verify. Used both as
/// a generic ES256 verifier (the interop test) and, at P7, for the phone's
/// own structural sanity-check of a daemon-sent Statement before signing it —
/// though that check has no matching key to verify a signature against, so
/// it validates JSON shape only, not via this function.
Uint8List? verifyEs256Compact(ECPublicKey pub, String compact) {
  final parts = compact.split('.');
  if (parts.length != 3) return null;

  final headerBytes = _decodeB64Url(parts[0]);
  final payload = _decodeB64Url(parts[1]);
  final sig = _decodeB64Url(parts[2]);
  if (headerBytes == null || payload == null || sig == null) return null;
  if (utf8.decode(headerBytes, allowMalformed: true) != _jwsHeader) return null;
  if (sig.length != 64) return null;

  final r = _readBigEndian(sig.sublist(0, 32));
  final s = _readBigEndian(sig.sublist(32, 64));

  final signingInput = '${parts[0]}.${parts[1]}';
  final digest = Uint8List.fromList(
    crypto.sha256.convert(utf8.encode(signingInput)).bytes,
  );

  final verifier = ECDSASigner(null, null)
    ..init(false, PublicKeyParameter(pub));
  final ok = verifier.verifySignature(digest, ECSignature(r, s));
  return ok ? payload : null;
}

File: internal/receipt/jws.go
// Package receipt implements signed receipts for permission decisions
// (MADR 0077): JWS ES256 signing/verification, the in-toto-style Statement
// shape, and append-only per-device chained storage.
package receipt

import (
	"crypto/ecdsa"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"math/big"
	"strings"
)

// jwsHeader is the fixed ES256 JWS Protected Header this package produces and
// accepts. No "kid": device/daemon identity is established out-of-band via
// the enrolled-cert lookup (MADR 0077 D9), not a JWS header field, which
// would be one more place to spoof.
const jwsHeader = `{"alg":"ES256"}`

// jwsHeaderB64 is the fixed base64url encoding of jwsHeader, computed once.
var jwsHeaderB64 = base64.RawURLEncoding.EncodeToString([]byte(jwsHeader))

// ErrMalformedJWS is returned by VerifyES256Compact when the input is not a
// well-formed three-part compact serialization.
var ErrMalformedJWS = errors.New("malformed JWS compact serialization")

// ErrWrongAlg is returned by VerifyES256Compact when the protected header is
// not exactly {"alg":"ES256"} — no algorithm negotiation, a wrong or foreign
// alg is rejected outright rather than downgraded to weaker validation.
var ErrWrongAlg = errors.New("JWS header is not exactly {\"alg\":\"ES256\"}")

// ErrSignatureInvalid is returned by VerifyES256Compact when the signature
// does not verify against the given public key.
var ErrSignatureInvalid = errors.New("JWS signature invalid")

// SignES256Compact signs payload as a JWS in Compact Serialization (RFC 7515
// §3.1) using ES256 (RFC 7518 §3.4): the ECDSA P-256 signature is the raw
// 64-byte R||S concatenation, big-endian, 32 bytes each — not ASN.1 DER.
func SignES256Compact(priv *ecdsa.PrivateKey, payload []byte) (string, error) {
	payloadB64 := base64.RawURLEncoding.EncodeToString(payload)
	signingInput := jwsHeaderB64 + "." + payloadB64
	sum := sha256.Sum256([]byte(signingInput))

	r, s, err := ecdsa.Sign(rand.Reader, priv, sum[:])
	if err != nil {
		return "", fmt.Errorf("sign: %w", err)
	}
	sig := make([]byte, 64)
	r.FillBytes(sig[:32])
	s.FillBytes(sig[32:])
	sigB64 := base64.RawURLEncoding.EncodeToString(sig)

	return signingInput + "." + sigB64, nil
}

// VerifyES256Compact verifies compact against pub and, on success, returns
// the decoded payload bytes. Rejects anything that is not exactly a
// three-part compact serialization with an {"alg":"ES256"} header.
func VerifyES256Compact(pub *ecdsa.PublicKey, compact string) ([]byte, error) {
	parts := strings.Split(compact, ".")
	if len(parts) != 3 {
		return nil, ErrMalformedJWS
	}
	headerB, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("%w: header: %v", ErrMalformedJWS, err)
	}
	if string(headerB) != jwsHeader {
		return nil, ErrWrongAlg
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("%w: payload: %v", ErrMalformedJWS, err)
	}
	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return nil, fmt.Errorf("%w: signature: %v", ErrMalformedJWS, err)
	}
	if len(sig) != 64 {
		return nil, fmt.Errorf("%w: signature is %d bytes, want 64", ErrMalformedJWS, len(sig))
	}
	r := new(big.Int).SetBytes(sig[:32])
	s := new(big.Int).SetBytes(sig[32:])

	signingInput := parts[0] + "." + parts[1]
	sum := sha256.Sum256([]byte(signingInput))
	if !ecdsa.Verify(pub, sum[:], r, s) {
		return nil, ErrSignatureInvalid
	}
	return payload, nil
}

// DecodePayloadUnverified returns compact's payload bytes without checking
// its signature. For display/inspection only (the CLI's `receipts show` on
// an entry whose signer key is unavailable, or a quick summary while
// listing) — never use this result for a trust decision; call
// VerifyES256Compact for that.
func DecodePayloadUnverified(compact string) ([]byte, error) {
	parts := strings.Split(compact, ".")
	if len(parts) != 3 {
		return nil, ErrMalformedJWS
	}
	headerB, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("%w: header: %v", ErrMalformedJWS, err)
	}
	if string(headerB) != jwsHeader {
		return nil, ErrWrongAlg
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("%w: payload: %v", ErrMalformedJWS, err)
	}
	return payload, nil
}
